### DISCUSSION TITLE
unnecessary_literal_unwrap gets me confused

### DISCUSSION THREAD

**Discussion by tshepang** [rust-lang/rust-clippy#11109] [Q&A] [answered]
I get the warning with this:

```rust
let version = option_env!("CARGO_PKG_VERSION");
let version = format!(
    "version: {}",
    version.unwrap_or_("(unknown version)"),
);
```

It gets quiet when I change it to this:

```rust
let version = option_env!("CARGO_PKG_VERSION");
let version = format!(
    "version: {}",
    if let Some(version) = version {
        version
    } else {
        "(unknown version)"
    }
};
```

Is this expected?

**Comment by y21** [comment]
The `option_env!` macro expands to `Some("...")` or `None`, so the clippy lint sees that code in its expanded form as:
```rs
let version = Some("<package version"); // option_env! expansion
let header = format!("{}",
  version.unwrap_or_("(unknown version)")
);
```
which makes it think that the `unwrap_or` is unnecessary because `version` is always `Some`.

This looks like a bug, and it should check if the `Some(...)`/`None` comes from an expansion and not lint in that case.

**Reply by tshepang** [reply]
as a sidenote, the lint goes quiet when `unwrap_or` happens on `option_env!`

**Reply by y21** [reply]
> as a sidenote, the lint goes quiet when `unwrap_or` happens on `option_env!`

Looking at the code now, it already does check if the receiver of the `unwrap_or` call comes from an expansion and not lint then (e.g. `option_env!("V").unwrap()`), so this seems to be why it correctly ignores it for that case but not if you store it in a variable.

**Reply by y21** [reply, marked as answer]
Submitted a PR that fixes this: #11110
(deleted & reposted it here as a reply to this answer and not a new one)
